### PR TITLE
Pause rolling-upgrade process of tidb statefulset

### PR DESCRIPTION
**Pull request by shuijing198799**
### What problem does this PR solve? 
Achieved the need to pause TiDB cluster upgrades

### What is changed and how it works?
Add a webhoo to hijack the request of update statefulset from apiserver, add annotation to crd tidb-cluster and pause the rolling-update process. And also provider a tkcli tool to detect the update information of tidb statefulset.

use tkcli to fetch the tidb upgrade information:

$ **./tkctl upinfo**
**Name:                  e2e-cluster1**
Namespace:             e2e-cluster1
**CreationTimestamp:     2019-05-06 20:28:28 +0800 CST**
**Statu:                 Normal**
  **Name                 State**
  **----                 -----**
  **e2e-cluster1-tidb-0  updated**
  **e2e-cluster1-tidb-1  updated**

use kubectl to change annotate to CRD tidb-cluster:

**kubectl annotate tc e2e-cluster1 -n e2e-cluster1 tidb.pingcap.com/tidb-partition=0 --overwrite
tidbcluster.pingcap.com/e2e-cluster1 annotated**

Tests 

 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Add yaml files
 - Has Go code change

@xiaojingchen @tennix @aylei @weekface PTAL

### Does this PR introduce a user-facing change?:
 
 ```release-note
NONE
 ```

### CODE REVIEW

File: manifests/webhook.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  template:
+    metadata:
+      namespace: pingcap
+      labels:
+        app: admission-webhook-example
+    spec:
+      serviceAccount: admission-webhook-example-sa
+      containers:
+        - name: admission-webhook-example
+          image: hub.pingcap.net/yinliang/pingcap/tidb-operator:latest
```
Comment (tennix) [inline manifests/webhook.yaml]: Don't use the private registry.

File: manifests/webhook.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: admission-webhook-example-svc
+  namespace: pingcap
```
Comment (tennix) [inline manifests/webhook.yaml]: Leave the namespace field to allow user specify it from the command line.

File: manifests/webhook.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  labels:
+    app: admission-webhook-example
```
Comment (tennix) [inline manifests/webhook.yaml]: Why name it `example`?

File: pkg/webhook/route/route.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+	deserializer := util.GetCodec()
+
+	// The AdmissionReview that will be returned
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		} else {
+			responseAdmissionReview.Response = util.ARFail(err)
+			goto returnData
```
Comment (tennix) [inline pkg/webhook/route/route.go]: Abstract this as a function to avoid using `goto`.

File: manifests/webhook.yaml (hunk 2 above)
Comment (tennix) [reply manifests/webhook.yaml]: Oh, seems `namespace` is required by cluster role binding. So you can keep it here, but please change it to `tidb-admin` the default namespace in our documentations.

File: manifests/webhook.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    spec:
+      serviceAccount: admission-webhook-example-sa
+      containers:
+        - name: admission-webhook-example
+          image: hub.pingcap.net/yinliang/pingcap/tidb-operator:latest
+          imagePullPolicy: Always
+          command: 
+          - /usr/local/bin/tidb-webhook
+          env:
+            - name: NAMESPACE
```
Comment (tennix) [inline manifests/webhook.yaml]: Please use downward API to get namespace dynamically.

File: manifests/webhook.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+              value: pingcap
+            - name: SERVICENAME
```
Comment (tennix) [inline manifests/webhook.yaml]: s/SERVICENAME/SERVICE_NAME/

File: cmd/webhook/main.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (xiaojingchen) [inline cmd/webhook/main.go]:
```suggestion
// Copyright 2019 PingCAP, Inc.
```

File: pkg/webhook/util/certs.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+type CertContext struct {
+	Cert        []byte
+	Key         []byte
+	SigningCert []byte
+}
+
+// Setup the server cert. For example, user apiservers and admission webhooks
+// can use the cert to prove their identify to the kube-apiserver
+func SetupServerCert(namespaceName, serviceName string) (*CertContext, error) {
```
Comment (tennix) [inline pkg/webhook/util/certs.go]: Why create certificates temporarily? Does this need to be persistent?

File: pkg/webhook/webhook.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	server := &http.Server{
+		Addr:      ":443",
+		TLSConfig: sCert,
+	}
+
+	return &WebhookServer{
+		KubeCli:    kubecli,
+		Cli:        cli,
+		Context:    context,
+		ConfigName: "validating-webhook-configuration",
```
Comment (tennix) [inline pkg/webhook/webhook.go]: This should be configurable.

File: pkg/webhook/webhook.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+	return ws.Server.ListenAndServeTLS("", "")
+}
+
+func (ws *WebhookServer) Shutdown() error {
+	return ws.Server.Shutdown(nil)
+}
+
+func strPtr(s string) *string { return &s }
+
+func (ws *WebhookServer) RegisterWebhook(namespace string, svcName string) error {
```
Comment (tennix) [inline pkg/webhook/webhook.go]: Should this be applied to the cluster with a yaml file?

**Comment by xiaojingchen** [comment]
@shuijing198799  webhook is a new component of tidb-operator, so need support helm to manage it.

File: pkg/webhook/statefulset/statefulset.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+	glog.Infof("admit statefulsets")
+
+	setResource := metav1.GroupVersionResource{Group: "apps", Version: "v1beta1", Resource: "statefulsets"}
+	if ar.Request.Resource != setResource {
+		err := fmt.Errorf("expect resource to be %s", setResource)
+		glog.Errorf("%v", err)
+		return util.ARFail(err)
+	}
+
+	cli, _, err := util.GetNewClient()
```
Comment (tennix) [inline pkg/webhook/statefulset/statefulset.go]: Create a new client every time a stateful set is updated?

File: pkg/webhook/statefulset/statefulset.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+	"k8s.io/api/admission/v1beta1"
+	apps "k8s.io/api/apps/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func AdmitStatefulSets(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	glog.Infof("admit statefulsets")
+
+	setResource := metav1.GroupVersionResource{Group: "apps", Version: "v1beta1", Resource: "statefulsets"}
+	if ar.Request.Resource != setResource {
```
Comment (tennix) [inline pkg/webhook/statefulset/statefulset.go]: Why compare like this? Does the webhook already only catch the stateful set `update` event?

File: pkg/webhook/statefulset/statefulset.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	namespace := ar.Request.Namespace
+
+	raw := ar.Request.OldObject.Raw
+	set := apps.StatefulSet{}
+	deserializer := util.GetCodec()
+	if _, _, err := deserializer.Decode(raw, nil, &set); err != nil {
+		glog.Error(err)
+		return util.ARFail(err)
+	}
+
```
Comment (xiaojingchen) [inline pkg/webhook/statefulset/statefulset.go]: here need to filter the statefulSet which is not created by `tidb-operator`.

File: pkg/tkctl/cmd/upinfo/upinfo.go [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+									}
+								}
+
+							} else {
+								state = WAITING
+							}
+						} else {
+							state = UPDATED
+						}
+						w.WriteLine(readable.LEVEL_1, "%s\t%s\t", pod.Name, state)
```
Comment (weekface) [inline pkg/tkctl/cmd/upinfo/upinfo.go]: Do we need to display the image before and after the upgrade?

File: cmd/webhook/main.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+	"syscall"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/pkg/webhook"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+)
+
+func main() {
+
+	cli, kubeCli, err := util.GetNewClient()
```
Comment (weekface) [inline cmd/webhook/main.go]: https://github.com/pingcap/tidb-operator/blob/617546b792be61e253eb3cc0152e953069120365/cmd/controller-manager/main.go#L89-L101

File: manifests/webhook-rbac.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: admission-webhook-example-rbac
+subjects:
+- kind: ServiceAccount
+  namespace: pingcap
+  name: admission-webhook-example-sa
```
Comment (weekface) [inline manifests/webhook-rbac.yaml]: `example `?

File: pkg/webhook/statefulset/statefulset.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	"k8s.io/api/admission/v1beta1"
+	apps "k8s.io/api/apps/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func AdmitStatefulSets(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	glog.Infof("admit statefulsets")
```
Comment (weekface) [inline pkg/webhook/statefulset/statefulset.go]: log more pieces of information, for example, the namespace and name of the statefulset

File: pkg/webhook/statefulset/statefulset.go [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+		return util.ARFail(err)
+	}
+
+	tc, err := cli.PingcapV1alpha1().TidbClusters(namespace).Get(set.Labels[label.InstanceLabelKey], metav1.GetOptions{})
+	if err != nil {
+		glog.Errorf("fail to fetch tidbcluster info namespace %s clustername(instance) %s err %v", namespace, set.Labels[label.InstanceLabelKey], err)
+		return util.ARFail(err)
+	}
+
+	if set.Labels[label.ComponentLabelKey] == "tidb" {
```
Comment (weekface) [inline pkg/webhook/statefulset/statefulset.go]:
```suggestion
	if set.Labels[label.ComponentLabelKey] == label.TiDBLabelVal {
```

File: cmd/webhook/main.go [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
```
[the remark this answers is not in the document]
Comment (shuijing198799) [reply cmd/webhook/main.go]: It seems that k8s does not provide such a admission controller。

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 14 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: It seems reasonable

File: cmd/webhook/main.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (weekface) [reply cmd/webhook/main.go]: webhook is not a nice component name.

File: cmd/webhook/main.go [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+func main() {
+
+	cli, kubeCli, err := util.GetNewClient()
+	if err != nil {
+		glog.Fatalf("failed to get client: %v", err)
+	}
+
+	ns := os.Getenv("NAMESPACE")
+	if ns == "" {
```
Comment (xiaojingchen) [inline cmd/webhook/main.go]: use `strings.TrimSpace(ns)== "" `

Comment (xiaojingchen) [inline cmd/webhook/main.go]: use `strings.TrimSpace(ns)== "" `

File: pkg/tkctl/cmd/upinfo/upinfo.go [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+)
+
+const (
+	upinfoLongDesc = `
+		Get tidb cluster component upgrade info.
+
+		You can omit --tidbcluster=<name> option by running 'tkc use <clusterName>',
+`
+	upinfoExample = `
+		# get current tidb cluster info (set by tkc use)
```
Comment (xiaojingchen) [inline pkg/tkctl/cmd/upinfo/upinfo.go]:
```suggestion
		# get current tidb cluster info (set by tkc user)
```

File: pkg/webhook/webhook.go [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+	Context *util.CertContext
+	// configuration name
+	ConfigName string
+	// http server
+	Server *http.Server
+}
+
+func NewWebHookServer(kubecli kubernetes.Interface, cli versioned.Interface, context *util.CertContext) *WebhookServer {
+
+	http.HandleFunc("/statefulsets", route.ServeStatefulSets)
```
Comment (xiaojingchen) [inline pkg/webhook/webhook.go]: make "/statefulsets" as a variable,

**Comment by cofyc** [comment]
How about moving the design proposal in google docs to the repo?

File: charts/tidb-operator/values.yaml [hunk 23]
```diff
@@ -55,3 +55,11 @@ scheduler:
...
...WVB6NW1sRmlmRW5FTEQwVkY0ak5pNFRWblBLYVhocEFCYgpNS3UwOXRuWXFnTHRacURBc0c4ZCtxWVlKOVZaQjMreUVnWTMwY043K1phVTlJcS81Q3R3N2p3d3JFMVEwTFN5CjhnQk9sZFVGUzVUMXFBOXhyOGx6WW1jdTl3TWpEZUNaTGxrdWdzWFlBRGF5eHRMRnpBU0kzWDBDZ1lFQXcrcTYKQnhGMjhkUFdwQk95bk5mcUIxdVc1Lyt6S3RhREh4Y1RMenRpaWdrM2NUbm1seTB0Zll4eExpeGs4aFVmV3dmQwoyejFDNmhxV1RFeS9nZTJHaWhlOFJDL1NFMGhvNEZyYUYwNVAvM0Voa0lwOFM1cWpobmdSdjJ2S1ZKU1BTLzNpCnJPRHBxNWxKU3pVc2lXWktTRFJ6RXlvRXI3Y2NodXpsNm8vYjFHTUNnWUFXMnE0bHFpRlR6YU1BZDFXZEZ6TkEKa3l4MUcwbmVTVm5UL0NCN2VCYklUT1FEZURjWlNEYnAzdVM2c3NWNE1oOG45QmFkUkw5TEFaWjJMZ1I4V2d1awo5dzRaT2hwd1NncEZHQnM2NUhTWDFLMERMMGV1R2k5aTlJZ2Jpek9Zb1dDYXBCaXhsY2RhOXE4UDRDWk1YdjBsClh6cUE0a3NSYWRrVGFXb0JkZjdlK1FLQmdRQ01PN0dZYVpuWUV5ZXFIaWRWT1A5YzhPUWRmOXVCS0RhR3hIK0kKNlZycEx4dXVRaUdaeUU2dGRSOTJTM1NlQ3h4OGpodWR6a1hkUjlhWkdyNWhnK0V2YXpucEhNbWozOHFoL0tYcQoxTFZTUkdvZ0NlQmFSWUFUT2xLSno5L3Rsc0RUUmhGQWFXVVkzUW1uZVpJOWJ5aHdjb3FFQnlsL2xQUW9aOTFTCmZsSjl2UUtCZ0dBTXRybTRFVkRTckF0QS9qeXg3NFB3SThjc2xtUEQ1d3JaWEZQVTUvc3FyRGNKbUNqSXBJRWUKRFdSaWNYL09aSUpRbDdjNkFOQjc1U1VUMHZ4dXJXeGp6WWlPVjRvUTJVUTBFc1VxcUh3RWxDVWZjclcxQnh2eQpmMnE3aGtUL0MySWwyRjhmRmxnYjJmZFVmaFM5MjZSbTF0aUUyMzJMU3MwYmZiWHArSDFmCi0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg==
```
Comment (tennix) [inline charts/tidb-operator/values.yaml]: Why put the key here? This should be filled by the users, the default value is unnecessary. Just leave some documents in the comment to tell user how to fill this value.

Comment (shuijing198799) [reply charts/tidb-operator/values.yaml]: done to leave the documents in comment as a guider

**Comment by shuijing198799** [comment]
@onlymellb @weekface PTAL

File: manifests/create-cert.sh [hunk 24]
```diff
@@ -0,0 +1,119 @@
+#!/bin/bash
+
+set -e
+
+usage() {
+    cat <<EOF
+Generate certificate suitable for use with an sidecar-injector webhook service.
```
Comment (weekface) [inline manifests/create-cert.sh]: `sidecar-injector `? 

Where does this script come from? Do we should give a link if it is a modified version.

File: manifests/create-cert.sh [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+
+This script uses k8s' CertificateSigningRequest API to a generate a
+certificate signed by k8s CA suitable for use with sidecar-injector webhook
```
Comment (weekface) [inline manifests/create-cert.sh]: ditto

File: manifests/create-cert.sh [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+services. This requires permissions to create and approve CSR. See
+https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster for
+detailed explantion and additional instructions.
+
+The server key/cert k8s CA cert are stored in a k8s secret.
```
Comment (weekface) [inline manifests/create-cert.sh]: You don't create the secret in this script.

File: charts/tidb-operator/values.yaml [hunk 27]
```diff
@@ -57,3 +57,16 @@ scheduler:
...
```
Comment (weekface) [inline charts/tidb-operator/values.yaml]: leave these keys empty.

File: pkg/tkctl/cmd/upinfo/upinfo.go [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+	if err != nil {
+		return err
+	}
+	setName := controller.TiDBMemberName(tc.Name)
+	set, err := o.KubeCli.AppsV1beta1().StatefulSets(o.Namespace).Get(setName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	podList, err := o.KubeCli.CoreV1().Pods(o.Namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s,%s=%s", label.InstanceLabelKey, tc.Name, label.ComponentLabelKey, "tidb"),
```
Comment (weekface) [inline pkg/tkctl/cmd/upinfo/upinfo.go]: use `label.New().Cluster().TiDB()....` instead.

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 14 above)
Comment (weekface) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: What is the display effect?

File: pkg/webhook/route/route.go [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+// function
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	var body []byte
+	var contentType string
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+	deserializer := util.GetCodec()
+
+	// The AdmissionReview that will be returned
+	if r.Body != nil {
```
Comment (weekface) [inline pkg/webhook/route/route.go]:
```
	if r.Body == nil {
		err := errors.New("request body is nil!")
		responseAdmissionReview.Response = util.ARFail(err)
		marshalAndWrite(responseAdmissionReview, w)
		return
        }

        if data, err := ioutil.ReadAll(r.Body); err != nil {
          responseAdmissionReview.Response = util.ARFail(err)
          marshalAndWrite(responseAdmissionReview, w)
          return
        }

        body = data
```

File: charts/tidb-operator/templates/admission-controller-deployment.yaml [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+        - name: tidb-admission-controller-pod
+          image: {{ .Values.operatorImage }}
+          imagePullPolicy: {{ .Values.imagePullPolicy | default "IfNotPresent" }}
+          command:
+            - /usr/local/bin/tidb-admission-controller
+            - -tlsCertFile=/etc/webhook/certs/cert.pem
+            - -tlsKeyFile=/etc/webhook/certs/key.pem
+            - -alsologtostderr
+            - -v={{ .Values.admissionController.logLevel }}
+            - 2>&1
```
Comment (tennix) [inline charts/tidb-operator/templates/admission-controller-deployment.yaml]: Is this necessary?

File: charts/tidb-operator/templates/admission-controller-deployment.yaml [hunk 31]
```diff
@@ -0,0 +1,101 @@
+{{- if .Values.admissionController.create }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: admission-controller-certs
+data:
+  cert.pem: {{ .Values.admissionController.cert }}
+  key.pem: {{ .Values.admissionController.key }}
```
Comment (onlymellb) [inline charts/tidb-operator/templates/admission-controller-deployment.yaml]: You can create cert's secret like [this](https://github.com/goharbor/harbor-helm/blob/master/templates/ingress/secret.yaml)

Comment (shuijing198799) [reply charts/tidb-operator/templates/admission-controller-deployment.yaml]: With this method, the key，cert and ca will be updated each time you upgrade.

File: charts/tidb-operator/templates/admission-controller-deployment.yaml (hunk 30 above)
Comment (shuijing198799) [reply charts/tidb-operator/templates/admission-controller-deployment.yaml]: you mean the args？

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 14 above)
Comment (weekface) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: show the output of the upinfo command like aylei does: https://github.com/pingcap/tidb-operator/pull/387#issue-269955716

File: pkg/tkctl/cmd/upinfo/upinfo.go [hunk 32]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+func renderTCUpgradeInfo(tc *v1alpha1.TidbCluster, set *apps.StatefulSet, podList *v1.PodList) (string, error) {
+	return readable.TabbedString(func(out io.Writer) error {
+		w := readable.NewPrefixWriter(out)
+		dbPhase := tc.Status.TiDB.Phase
+		w.WriteLine(readable.LEVEL_0, "Name:\t%s", tc.Name)
+		w.WriteLine(readable.LEVEL_0, "Namespace:\t%s", tc.Namespace)
+		w.WriteLine(readable.LEVEL_0, "CreationTimestamp:\t%s", tc.CreationTimestamp)
+		w.WriteLine(readable.LEVEL_0, "Statu:\t%s", dbPhase)
```
Comment (weekface) [inline pkg/tkctl/cmd/upinfo/upinfo.go]:
```suggestion
		w.WriteLine(readable.LEVEL_0, "Status:\t%s", dbPhase)
```

**Comment by weekface** [comment]
ref: https://github.com/kubernetes/kubernetes/issues/72944

Allow Mutating|ValidatingWebhookConfiguration to use secret for CABundle

File: pkg/webhook/route/route.go [hunk 33]
```diff
@@ -0,0 +1,95 @@
...
+	if contentType != "application/json" {
+		err := errors.New("expect application/json")
+		responseAdmissionReview.Response = util.ARFail(err)
+		marshalAndWrite(responseAdmissionReview, w)
+		return
+	}
+
+	// The AdmissionReview that was sent to the webhook
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		responseAdmissionReview.Response = util.ARFail(err)
```
Comment (weekface) [inline pkg/webhook/route/route.go]:
```suggestion
		responseAdmissionReview.Response = util.ARFail(err)
		marshalAndWrite(responseAdmissionReview, w)
		return
```
?

File: tests/actions.go [hunk 34]
```diff
@@ -1106,8 +1035,26 @@ func (oa *operatorActions) tidbMembersReadyFn(tc *v1alpha1.TidbCluster) (bool, e
 			ns, tidbSetName, tidbSet.Status.ReadyReplicas, tidbSet.Status.Replicas)
 		return false, nil
 	}
-	if c, ok := getComponentContainer(tidbSet); !ok || tc.Spec.TiDB.Image != c.Image {
-		glog.Infof("statefulset: %s/%s .spec.template.spec.containers[name=tikv].image(%s) != %s",
+
+	if upgradePaused() {
+
+		time.Sleep(5 * time.Minute)
```
Comment (weekface) [inline tests/actions.go]: The upgrade process take more than 5 minutes, so you must sleep more than it, for exmaple 10 minutes.

**Comment by weekface** [comment]
add a release note like: https://github.com/pingcap/tidb-operator/pull/477

File: charts/tidb-operator/templates/admission-controller-deployment.yaml [hunk 35]
```diff
@@ -0,0 +1,101 @@
...
+    targetPort: 443
+  selector:
+    app.kubernetes.io/name: {{ template "chart.name" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/component: admission-controller
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: tidb-admission-controller-deployment
```
Comment (aylei) [inline charts/tidb-operator/templates/admission-controller-deployment.yaml]: I don't want to argue about naming style, bug add `-deployment` suffix to a deployment object is wired, so does the `tidb-admission-controller-pod` and `admission-controller-svc` and `admission-controller-cfg`

File: pkg/webhook/statefulset/statefulset.go [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+	glog.Infof("admit statefulsets [%s/%s]", name, namespace)
+
+	setResource := metav1.GroupVersionResource{Group: "apps", Version: "v1beta1", Resource: "statefulsets"}
+	if ar.Request.Resource != setResource {
+		err := fmt.Errorf("expect resource to be %s", setResource)
+		glog.Errorf("%v", err)
+		return util.ARFail(err)
+	}
+
+	if versionCli == nil {
```
Comment (aylei) [inline pkg/webhook/statefulset/statefulset.go]: Could we put the initialization of `versionCli` to `init()` in favor of fail-fast

File: pkg/webhook/util/util.go [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+package util
+
+import (
+	"crypto/tls"
+
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// toAdmissionResponse is a helper function to create an AdmissionResponse
```
Comment (aylei) [inline pkg/webhook/util/util.go]:
```suggestion
// ARFail is a helper function to create an AdmissionResponse
```

File: pkg/webhook/util/util.go [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+	return &v1beta1.AdmissionResponse{
+		Allowed: false,
+		Result: &metav1.Status{
+			Message: err.Error(),
+			Reason:  metav1.StatusReasonNotAcceptable,
+		},
+	}
+}
+
+// toAdmissionResponse return allow to action
```
Comment (aylei) [inline pkg/webhook/util/util.go]:
```suggestion
// ARSuccess return allow to action
```

File: tests/actions.go [hunk 39]
```diff
@@ -792,157 +817,29 @@ func (oa *operatorActions) UpgradeTidbClusterOrDie(info *TidbClusterConfig) {
...
+	}
+	if len(pod.Spec.Containers) == 0 {
+		glog.Errorf("no container in this pod [%s/%s]", namespace, name)
+		return nil, false
+	}
+
+	for _, container := range pod.Spec.Containers {
+		if container.Name == label.TiDBLabelVal ||
+			container.Name == label.TiKVLabelVal ||
+			container.Name == label.PDLabelVal {
```
Comment (aylei) [inline tests/actions.go]:
```suggestion
		if container.Name == v1alpha1.PDMemberType.String() ||
```
Though the literal value is same coincidentally, we should use `v1alpha1.XXMemberType` here (So do `TiDB` and `TiKV`).

File: tests/actions.go [hunk 40]
```diff
@@ -792,157 +817,29 @@ func (oa *operatorActions) UpgradeTidbClusterOrDie(info *TidbClusterConfig) {
...
-
 func (oa *operatorActions) DeployMonitor(info *TidbClusterConfig) error { return nil }
 func (oa *operatorActions) CleanMonitor(info *TidbClusterConfig) error  { return nil }
 
-func getComponentContainer(set *v1beta1.StatefulSet) (corev1.Container, bool) {
-	name := set.Labels[label.ComponentLabelKey]
-	for _, c := range set.Spec.Template.Spec.Containers {
-		if c.Name == name {
-			return c, true
+func getPodContainer(kubeCli kubernetes.Interface, namespace string, memberName string) (*corev1.Container, bool) {
```
Comment (aylei) [inline tests/actions.go]: This function name is misleading, it retrieve the member container from member pod actually, so I think `getMemberContainer` is more descriptive

File: pkg/tkctl/cmd/upinfo/upinfo.go [hunk 41]
```diff
@@ -0,0 +1,213 @@
...
+		}
+		{
+			w.WriteLine(readable.LEVEL_1, "Name\tState\t")
+			w.WriteLine(readable.LEVEL_1, "----\t-----\t")
+			{
+				updateReplicas := set.Spec.UpdateStrategy.RollingUpdate.Partition
+
+				if len(podList.Items) != 0 {
+					for _, pod := range podList.Items {
+						var state string
```
Comment (aylei) [inline pkg/tkctl/cmd/upinfo/upinfo.go]: The logic inside the `for` loop is nested deeply, how about abstract this as a function?

File: pkg/webhook/route/route.go (hunk 29 above)
Comment (shuijing198799) [reply pkg/webhook/route/route.go]: data, err := ioutil.ReadAll(r.Body)
        if err != nil {
          responseAdmissionReview.Response = util.ARFail(err)
          marshalAndWrite(responseAdmissionReview, w)
          return
        }

        body = data

       done like this

File: pkg/webhook/route/route.go (hunk 33 above)
Comment (shuijing198799) [reply pkg/webhook/route/route.go]: maybe a space

File: pkg/webhook/util/util.go (hunk 38 above)
Comment (shuijing198799) [reply pkg/webhook/util/util.go]: done

File: tests/actions.go (hunk 40 above)
Comment (shuijing198799) [reply tests/actions.go]: done

File: tests/actions.go (hunk 39 above)
Comment (shuijing198799) [reply tests/actions.go]: done

File: pkg/webhook/statefulset/statefulset.go (hunk 36 above)
Comment (shuijing198799) [reply pkg/webhook/statefulset/statefulset.go]: I think it is not neccessry

File: tests/actions.go (hunk 34 above)
Comment (shuijing198799) [reply tests/actions.go]: upgradePause make sure that the last TiDB pod is upgraded and healthy, So I think 5 minute is enough.

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 14 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: I've already show it in <What is changed and how it works?> and I bold it.

File: charts/tidb-operator/templates/admission-controller-deployment.yaml (hunk 30 above)
Comment (shuijing198799) [reply charts/tidb-operator/templates/admission-controller-deployment.yaml]: It's necessary if the program did not init the log, and I change the method. done.

Comment (tennix) [reply charts/tidb-operator/templates/admission-controller-deployment.yaml]: So it's unnecessary now?

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 32 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: done

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 28 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: done

File: manifests/create-cert.sh (hunk 26 above)
Comment (shuijing198799) [reply manifests/create-cert.sh]: now I have

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 28 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: done

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 32 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: done

File: pkg/tkctl/cmd/upinfo/upinfo.go (hunk 41 above)
Comment (shuijing198799) [reply pkg/tkctl/cmd/upinfo/upinfo.go]: It is not complicate so I think it is not necessary。

File: pkg/webhook/util/util.go (hunk 37 above)
Comment (shuijing198799) [reply pkg/webhook/util/util.go]: done

File: charts/tidb-operator/templates/admission-controller-deployment.yaml (hunk 30 above)
Comment (shuijing198799) [reply charts/tidb-operator/templates/admission-controller-deployment.yaml]: It's not necessary now , I have remove it in yaml file.

**Pull request merged**